File: go/oracle/proposal_access.py
"""Capture proposal input validation, path policy and record visibility."""

from __future__ import annotations

import importlib
from dataclasses import asdict
from types import SimpleNamespace
from typing import Any


def fixtures() -> dict[str, Any]:
    service: Any = importlib.import_module("memento.service")
    authz: Any = importlib.import_module("memento.authz")
    runtime = object.__new__(service.MemoryService)
    bases: list[dict[str, Any]] = [
        {
            "kind": "create",
            "path": "/public/a.md",
            "concept_type": "anything",
            "title": "",
            "body": "body\r\n",
        },
        {"kind": "patch", "path": "/public/a.md"},
        {"kind": "rename", "path": "/public/a.md", "new_path": "/public/b.md"},
        {"kind": "trash", "path": "/public/a.md"},
        {
            "kind": "attach_asset_pack",
            "path": "/public/a.md",
            "asset_kind": "anything",
            "version": "unvalidated",
            "asset_id": "id",
            "zip_sha256": "unchecked",
            "manifest": {"nested": [1, None, True]},
        },
    ]
    inputs: list[list[Any]] = [
        [],
        [{}],
        [{"kind": "unknown"}],
        [{"kind": None}],
        [{"kind": 1}],
        [None],
    ]
    value: Any
    status: Any
    for base in bases:
        inputs.append([base])
        inputs.append([{**base, "extra": "rejected"}])
        for field in base:
            inputs.append([{key: value for key, value in base.items() if key != field}])
            for value in [None, 1, True, [], {}]:
                inputs.append([{**base, field: value}])
    for base in bases[:2]:
        for field in ["tags", "aliases"]:
            for value in [None, [], ["", "a", "a", "é"], [1], "x", {}]:
                inputs.append([{**base, field: value}])
        for field in ["description"] + (["title", "body"] if base["kind"] == "patch" else []):
            for value in [None, "", "text", 1, []]:
                inputs.append([{**base, field: value}])
    for status in [None, "active", "deprecated", "tombstone", "archived", 1, {}]:
        inputs.append([{**bases[1], "status": status}])
    normalization = []
    for changes in inputs:
        item: dict[str, Any] = {"changes": changes}
        try:
            item["expected"] = [
                change.model_dump(mode="json") for change in runtime._normalize_changes(changes)
            ]
        except Exception as exc:
            item["error_type"] = type(exc).__name__
            if isinstance(exc, service.ServiceError):
                item["error"] = str(exc)
        normalization.append(item)

    policies = [
        authz.EffectivePolicy(
            "author", ("proposer",), ("/public/",), ("/public/",), ("/private/",)
        ),
        authz.EffectivePolicy("author", ("reader",), ("/public/",), (), ("/private/",)),
        authz.EffectivePolicy("other", ("proposer",), ("/",), ("/",), ("/private/",)),
        authz.EffectivePolicy("other", ("curator",), ("/",), ("/",), ("/private/",)),
        authz.EffectivePolicy("other", ("admin",), ("/",), ("/",), ("/private/",)),
        authz.EffectivePolicy("other", ("curator", "admin"), ("/",), ("/",), ("/private/",)),
        authz.EffectivePolicy(
            "other", ("curator",), ("/public/", "/private/"), ("/public/",), ("/private/",)
        ),
    ]
    batches: list[list[dict[str, Any]]] = [[], *[[base] for base in bases], bases[:2]]
    for path in [
        "/private/a.md",
        "/trash/public/a.md",
        "/a.MD",
        "relative.md",
        "/public/../a.md",
        "/public/a.txt",
    ]:
        batches.extend([[{**bases[1], "path": path}], [{**bases[2], "new_path": path}]])
    batches.extend(
        [
            [bases[3], bases[1]],
            [bases[3], bases[3]],
            [{"kind": "trash", "path": f"/public/{i}.md"} for i in range(20)],
            [{"kind": "trash", "path": f"/public/{i}.md"} for i in range(21)],
        ]
    )
    authorization = []
    visibility = []
    for policy in policies:
        for changes in batches:
            for action in ["read", "write", "review"]:
                item = {"policy": asdict(policy), "changes": changes, "action": action}
                try:
                    runtime._validate_change_auth(
                        policy, runtime._normalize_changes(changes), action=action
                    )
                    item["allowed"] = True
                except (service.ServiceError, authz.AuthorizationError) as exc:
                    item["error"] = str(exc)
                authorization.append(item)
        for changes in batches + [[{"kind": "unknown"}], [{"kind": "patch"}]]:
            for require_write in [False, True]:
                proposal = SimpleNamespace(
                    author_principal="author", proposal_id="proposal", patch={"changes": changes}
                )
                item = {
                    "policy": asdict(policy),
                    "changes": changes,
                    "require_write": require_write,
                }
                try:
                    item["allowed"] = runtime._can_access_proposal(
                        policy, proposal, require_write=require_write
                    )
                    try:
                        runtime._require_proposal_access(
                            policy, proposal, require_write=require_write
                        )
                    except service.ForbiddenError as exc:
                        item["error"] = str(exc)
                except Exception as exc:
                    item["error_type"] = type(exc).__name__
                visibility.append(item)
    return {
        "normalization": normalization,
        "authorization": authorization,
        "visibility": visibility,
    }

File: internal/service/proposal_access.go
package service

import (
	"context"
	"errors"
	"fmt"
	"strings"

	"github.com/rcarmo/memento/internal/access"
	"github.com/rcarmo/memento/internal/control"
)

// Error is a service policy failure, distinct from malformed persisted input.
type Error struct {
	Class   string
	Message string
}

func (e *Error) Error() string { return e.Message }

// ChangeValidationError corresponds to source model-validation failures. These
// propagate from proposal visibility rather than being converted into a denial.
// Diagnostic text is not yet Pydantic-compatible.
type ChangeValidationError struct{ Message string }

func (e *ChangeValidationError) Error() string { return e.Message }

// ProposalChange contains the JSON-mode values of a validated change. Construct
// changes with NormalizeProposalChanges before passing them to policy helpers.
type ProposalChange map[string]any

// NormalizeProposalChanges preserves the source's defaults, nulls and list order.
// It validates shape only: path, concept type, assets and body policy are checked
// later by the service operations that consume these changes.
func NormalizeProposalChanges(values []any) ([]ProposalChange, error) {
	result := make([]ProposalChange, 0, len(values))
	for _, value := range values {
		raw, ok := value.(map[string]any)
		if !ok {
			return nil, &ChangeValidationError{"proposal change must be an object"}
		}
		kind, _ := raw["kind"].(string)
		required := []string{"kind", "path"}
		optional := []string{}
		lists := []string{}
		manifest := false
		switch kind {
		case "create":
			required = append(required, "concept_type", "title", "body")
			optional = []string{"description"}
			lists = []string{"tags", "aliases"}
		case "patch":
			optional = []string{"title", "description", "body", "status"}
			lists = []string{"tags", "aliases"}
		case "rename":
			required = append(required, "new_path")
		case "trash":
		case "attach_asset_pack":
			required = append(required, "asset_kind", "version", "asset_id", "zip_sha256")
			manifest = true
		default:
			label := fmt.Sprint(raw["kind"])
			if raw["kind"] == nil {
				label = "None"
			}
			return nil, &Error{"validation_error", "unsupported change kind: " + label}
		}
		change := ProposalChange{}
		for _, field := range required {
			v, ok := raw[field].(string)
			if !ok {
				return nil, &ChangeValidationError{field + " must be a string"}
			}
			change[field] = v
		}
		for _, field := range optional {
			v := raw[field]
			if v != nil {
				if _, ok := v.(string); !ok {
					return nil, &ChangeValidationError{field + " must be a string or null"}
				}
			}
			change[field] = v
		}
		if value := change["status"]; value != nil && value != "active" && value != "deprecated" && value != "tombstone" {
			return nil, &ChangeValidationError{"invalid concept status"}
		}
		for _, field := range lists {
			v, exists := raw[field]
			if v == nil && kind == "patch" {
				change[field] = nil
				continue
			}
			if !exists {
				change[field] = []string{}
				continue
			}
			items, ok := v.([]any)
			if !ok {
				return nil, &ChangeValidationError{field + " must be a string list"}
			}
			normalized := make([]string, 0, len(items))
			for _, item := range items {
				str, ok := item.(string)
				if !ok {
					return nil, &ChangeValidationError{field + " entries must be strings"}
				}
				normalized = append(normalized, str)
			}
			change[field] = normalized
		}
		if manifest {
			v, ok := raw["manifest"].(map[string]any)
			if !ok {
				return nil, &ChangeValidationError{"manifest must be an object"}
			}
			change["manifest"] = v
		}
		for field := range raw {
			if _, ok := change[field]; !ok {
				return nil, &ChangeValidationError{"unknown change field: " + field}
			}
		}
		result = append(result, change)
	}
	return result, nil
}

func validateArchivalBatch(changes []ProposalChange) error {
	trash := 0
	seen := map[string]bool{}
	for _, change := range changes {
		if change["kind"] == "trash" {
			trash++
		}
		seen[change["path"].(string)] = true
	}
	if trash > 0 {
		if len(changes) > 20 || trash != len(changes) {
			return &Error{"validation_error", "archival proposals require 1-20 trash changes without mixed mutations"}
		}
		if len(seen) != len(changes) {
			return &Error{"validation_error", "duplicate archival target"}
		}
	}
	return nil
}

// ValidateChangeAuthorization requires already normalised changes. As in the
// source, any action other than read/write requests both permissions.
func ValidateChangeAuthorization(policy access.EffectivePolicy, changes []ProposalChange, action string) error {
	if err := validateArchivalBatch(changes); err != nil {
		return err
	}
	for _, change := range changes {
		paths := []string{change["path"].(string)}
		if change["kind"] == "rename" {
			paths = append(paths, change["new_path"].(string))
		}
		for _, path := range paths {
			if strings.HasPrefix(path, "/trash/") {
				return &Error{"validation_error", "use trash, restore or purge for items in /trash/"}
			}
		}
		for _, path := range paths {
			if !strings.HasSuffix(path, ".md") {
				return &Error{"validation_error", "concept paths must end with .md"}
			}
		}
		actions := []string{action}
		if action != "read" && action != "write" {
			actions = []string{"read", "write"}
		}
		for _, action := range actions {
			for _, path := range paths {
				if _, err := access.AuthorizePath(policy, path, action); err != nil {
					return err
				}
			}
		}
	}
	return nil
}

// CanAccessProposal checks both author/curator ownership and every changed path.
// Admin alone does not bypass ownership, and review additionally requires writes.
func CanAccessProposal(policy access.EffectivePolicy, proposal control.ProposalRecord, requireWrite bool) (bool, error) {
	curator := false
	for _, role := range policy.Roles {
		if role == "curator" {
			curator = true
			break
		}
	}
	if proposal.AuthorPrincipal != policy.Principal && !curator {
		return false, nil
	}
	changes, err := proposalChanges(proposal)
	if err == nil {
		err = ValidateChangeAuthorization(policy, changes, "read")
	}
	if err == nil && requireWrite {
		err = ValidateChangeAuthorization(policy, changes, "write")
	}
	if err != nil {
		var serviceErr *Error
		var authErr *access.AuthorizationError
		if errors.As(err, &serviceErr) || errors.As(err, &authErr) {
			return false, nil
		}
		return false, err
	}
	return true, nil
}

func RequireProposalAccess(policy access.EffectivePolicy, proposal control.ProposalRecord, requireWrite bool) error {
	allowed, err := CanAccessProposal(policy, proposal, requireWrite)
	if err != nil {
		return err
	}
	if allowed {
		return nil
	}
	action := "read"
	if requireWrite {
		action = "review"
	}
	return &Error{"forbidden", fmt.Sprintf("principal %s cannot %s proposal %s", policy.Principal, action, proposal.ProposalID)}
}

func (q ProposalQueue) VisibleProposal(ctx context.Context, policy access.EffectivePolicy, id string) (control.ProposalRecord, error) {
	proposal, err := q.Proposals.Get(ctx, id)
	if err != nil {
		return control.ProposalRecord{}, err
	}
	if err = RequireProposalAccess(policy, proposal, false); err != nil {
		return control.ProposalRecord{}, err
	}
	return proposal, nil
}
